File: web/src/types/aiStatistics.ts
export interface AIFeatureStatistics {
  feature: string
  requests: number
  successful: number
  failed: number
  known_usage_requests: number
  unknown_usage_requests: number
  prompt_tokens: number
  completion_tokens: number
  total_tokens: number
  not_sent: number
  ambiguous: number
  deadline_failures: number
  timeout_failures: number
  truncated_outputs: number
  budget_rejections: number
  latency_samples: number
  p50_latency_ms: number | null
  p95_latency_ms: number | null
  p99_latency_ms: number | null
}

export interface AIStatisticsResponse {
  since: string
  until: string
  days: number
  features: AIFeatureStatistics[]
  queues: Array<{ feature: string; queued: number; running: number; oldest_queued_at: string | null; oldest_running_at: string | null }>
  provider_retry_attempts: number
  recovered_pre_io_failures: number
  latency_histogram: Record<string, number>
}

File: backend/app/schemas/ai_statistics.py
from datetime import datetime

from pydantic import BaseModel, Field


class AIFeatureStatistics(BaseModel):
    feature: str
    requests: int
    successful: int
    failed: int
    known_usage_requests: int
    unknown_usage_requests: int
    prompt_tokens: int
    completion_tokens: int
    total_tokens: int
    not_sent: int
    ambiguous: int
    deadline_failures: int
    timeout_failures: int
    truncated_outputs: int
    budget_rejections: int
    latency_samples: int
    p50_latency_ms: float | None
    p95_latency_ms: float | None
    p99_latency_ms: float | None


class AIQueueStatistics(BaseModel):
    feature: str
    queued: int
    running: int
    oldest_queued_at: datetime | None
    oldest_running_at: datetime | None


class AIStatisticsResponse(BaseModel):
    since: datetime
    until: datetime
    days: int
    features: list[AIFeatureStatistics]
    queues: list[AIQueueStatistics]
    provider_retry_attempts: int
    recovered_pre_io_failures: int
    latency_histogram: dict[str, int] = Field(default_factory=dict)
